src/c2ffi/generator.lisp

```diff
@@ -0,0 +1,755 @@
+target package."
+  (check-type c2ffi-spec-file (or pathname string))
+  (macrolet ((@ (var)
+                 `(setf ,var (compile-rules ,var))))
+    (@ include-sources)
+    (@ exclude-sources)
+    (@ include-definitions)
+    (@ exclude-definitions))
+  (with-standard-io-syntax
+    (with-input-from-file (in c2ffi-spec-file :external-format (asdf/driver:encoding-external-format :utf-8))
+      (with-output-to-file (*c2ffi-output-stream* output :if-exists :supersede
+                            :external-format (asdf/driver:encoding-external-format output-encoding))
+        (let* ((*package* (or (find-package package-name)
+                              (make-package package-name)))
+               ;; Make sure we use an uninterned symbol, so that it's neutral to READTABLE-CASE.
+               (package-name (make-symbol (package-name *package*)))
+               ;; Let's rebind a copy, so that when we are done with the generation
+               ;; then the CFFI type repository is also reverted back to the
+               ;; previous state. This avoids future redefinition warning if they
+               ;; get added to defctype and friends.
+               (cffi::*type-parsers* (copy-hash-table cffi::*type-parsers*))
+               (*anon-name-counter* 0)
+               (*anon-entities* (make-hash-table))
+               (*generated-names* (mapcar (lambda (key)
+                                            `(,key . ,(make-hash-table :test 'equal)))
+                                          +name-kinds+))
+               (*allow-pointer-type-simplification* allow-pointer-type-simplification)
+               (*allow-skipping-struct-fields* allow-skipping-struct-fields)
+               (*assume-struct-by-value-support* assume-struct-by-value-support)
+               (*ffi-name-transformer* (canonicalize-transformer-hook ffi-name-transformer))
+               (*ffi-type-transformer* (canonicalize-transformer-hook ffi-type-transformer))
+               (json (json:decode-json in)))
+          (output/string +generated-file-header+)
+          ;; some forms that are always emitted
+          (mapcar 'output/code `((defpackage ,package-name)
+                                 (in-package ,package-name)
+                                 (cffi:defctype ,(function-pointer-type-name) :pointer)))
+          (when (and foreign-library-name
+                     foreign-library-spec)
+            (when (stringp foreign-library-name)
+              (setf foreign-library-name (safe-read-from-string foreign-library-name)))
+            (output/code `(cffi:define-foreign-library ,foreign-library-name
+                            ,@foreign-library-spec))
+            ;; NOTE: maybe we should register this as an UIOP:REGISTER-IMAGE-RESTORE-HOOK?
+            ;; On some platforms DLL's need to be relinked after an image save and restart.
+            (output/code `(cffi:use-foreign-library ,foreign-library-name)))
```

Comment:
`define-foreign-library` registers the library, and it's possible to list all defined libraries, reopen them all, things like that. Right now, users will have to deal with that themselves, but the tools are there. If we're going to make it more automatic, then CFFI should be responsible for it, not CFFI/C2FFI, I suppose.